File: src/commands/inspect.ts
import { openDb } from "../store.ts";

interface InspectOptions {
  agent?: string;
  type?: string;
  session?: string;
  since?: string;
  limit?: number;
}

interface EventRow {
  id: string;
  ts: string;
  event_type: string;
  agent_slug: string;
  session_id: string | null;
  cwd: string | null;
  file_path: string | null;
  rel_file_path: string | null;
  payload: string | null;
}

export async function runInspect(opts: InspectOptions): Promise<void> {
  const db = openDb();

  const where: string[] = [];
  const params: Array<string | number> = [];
  if (opts.agent) {
    where.push("agent_slug = ?");
    params.push(opts.agent);
  }
  if (opts.type) {
    where.push("event_type = ?");
    params.push(opts.type);
  }
  if (opts.session) {
    where.push("session_id = ?");
    params.push(opts.session);
  }
  if (opts.since) {
    where.push("ts >= ?");
    params.push(opts.since);
  }
  const limit = Math.max(1, Math.min(opts.limit ?? 10, 1000));

  const sql = `
    SELECT id, ts, event_type, agent_slug, session_id, cwd, file_path, rel_file_path, payload
    FROM events
    ${where.length ? `WHERE ${where.join(" AND ")}` : ""}
    ORDER BY ts DESC
    LIMIT ${limit}
  `;
  const rows = db.query(sql).all(...params) as EventRow[];

  if (rows.length === 0) {
    console.log("(no matching events)");
    return;
  }

  for (const r of rows) {
    let payload: unknown = r.payload;
    if (r.payload) {
      try {
        payload = JSON.parse(r.payload);
      } catch {
        // leave as raw string
      }
    }
    const block = {
      ts: r.ts,
      agent: r.agent_slug,
      event_type: r.event_type,
      session_id: r.session_id,
      cwd: r.cwd,
      file_path: r.file_path,
      rel_file_path: r.rel_file_path,
      payload,
    };
    console.log(JSON.stringify(block, null, 2));
    console.log("---");
  }

  console.log(`(${rows.length} event${rows.length === 1 ? "" : "s"})`);
}

File: internal/commands/inspect.go
package commands

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/bilanc/posthook/internal/store"

	"github.com/spf13/cobra"
)

func newInspectCmd() *cobra.Command {
	var (
		agent   string
		evType  string
		session string
		since   string
		limit   int
	)
	cmd := &cobra.Command{
		Use:   "inspect",
		Short: "Print recent event payloads (default 10)",
		RunE: func(cmd *cobra.Command, args []string) error {
			return runInspect(agent, evType, session, since, limit)
		},
	}
	cmd.Flags().StringVar(&agent, "agent", "", "Filter by agent slug")
	cmd.Flags().StringVar(&evType, "type", "", "Filter by event_type")
	cmd.Flags().StringVar(&session, "session", "", "Filter by session_id")
	cmd.Flags().StringVar(&since, "since", "", "Only events with ts >= ISO timestamp")
	cmd.Flags().IntVar(&limit, "limit", 10, "Maximum number of events to print")
	return cmd
}

func runInspect(agent, evType, session, since string, limit int) error {
	db, err := store.Open()
	if err != nil {
		return err
	}
	if limit < 1 {
		limit = 1
	}
	if limit > 1000 {
		limit = 1000
	}

	var where []string
	var params []any
	if agent != "" {
		where = append(where, "agent_slug = ?")
		params = append(params, agent)
	}
	if evType != "" {
		where = append(where, "event_type = ?")
		params = append(params, evType)
	}
	if session != "" {
		where = append(where, "session_id = ?")
		params = append(params, session)
	}
	if since != "" {
		where = append(where, "ts >= ?")
		params = append(params, since)
	}
	whereClause := ""
	if len(where) > 0 {
		whereClause = "WHERE " + strings.Join(where, " AND ")
	}
	q := fmt.Sprintf(`
		SELECT id, ts, event_type, agent_slug, session_id, cwd, file_path, rel_file_path, payload
		FROM events
		%s
		ORDER BY ts DESC
		LIMIT %d`, whereClause, limit)

	rows, err := db.Query(q, params...)
	if err != nil {
		return err
	}
	defer rows.Close()

	count := 0
	for rows.Next() {
		var id, ts, eventType, agentSlug string
		var sessionID, cwd, filePath, relFilePath, payload sql.NullString
		if err := rows.Scan(&id, &ts, &eventType, &agentSlug, &sessionID, &cwd, &filePath, &relFilePath, &payload); err != nil {
			return err
		}
		block := map[string]any{
			"ts":            ts,
			"agent":         agentSlug,
			"event_type":    eventType,
			"session_id":    nullToAny(sessionID),
			"cwd":           nullToAny(cwd),
			"file_path":     nullToAny(filePath),
			"rel_file_path": nullToAny(relFilePath),
		}
		if payload.Valid {
			var parsed any
			if err := json.Unmarshal([]byte(payload.String), &parsed); err == nil {
				block["payload"] = parsed
			} else {
				block["payload"] = payload.String
			}
		} else {
			block["payload"] = nil
		}
		data, _ := json.MarshalIndent(block, "", "  ")
		fmt.Println(string(data))
		fmt.Println("---")
		count++
	}
	if err := rows.Err(); err != nil {
		return err
	}
	if count == 0 {
		fmt.Println("(no matching events)")
		return nil
	}
	plural := "s"
	if count == 1 {
		plural = ""
	}
	fmt.Printf("(%d event%s)\n", count, plural)
	return nil
}

func nullToAny(s sql.NullString) any {
	if !s.Valid {
		return nil
	}
	return s.String
}
